### PR TITLE
Python Scoring Example

### PR DESCRIPTION
**Pull request by mtanco**
Guide for step by step setting up a HTTP server for predictions and shapley values in ubuntu

### CODE REVIEW

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  , "result": [-0.31093353033065796
+               , -1.2875813245773315
+               , -0.4879034757614136
+               , -1.1715366840362549
+               , -0.6523758769035339
+               , 0.0]
+}
+```
+
+
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: ultra-nit: please remove the extra trailing lines (I tend to keep only a single EOL at EOF to reduce diff noise).

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
+# Shapley Values in ec2
+
+This guide walks through setting up an http server in an ec2 instance which can return both predictions and shapley values.
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: nit: Consider using HTTP (=all caps) consistently throughout the text. Maybe also EC2.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+This guide walks through setting up an http server in an ec2 instance which can return both predictions and shapley values. 
+
+1. Example ec2 Environment Setup
+2. Install the requirements
+3. Run the HTTP Server
+
+## Example ec2 Environment Setup 
+
+The linux commands in this example are for Ubuntu, but any flavor of linux would work fine.
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: I wouldn't be so optimistic tbh :). Not all distros have `apt` and `apt-get`, which is used below. Probably Debian-based distros are likely to have these, but Debian != Linux.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+**Open Ports:** SSH 22, Custom TCP 9090
+
+**Notes:** Enable public IPv4
+
+### Connect to the System
+
+Example command to ssh to the system with a private key
+
+```bash
+chmod 600 ***key***.pem
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Not sure I would make this part of the steps :thinking: (i.e., forcing people to change access rights to their certificates is a bit weird). I would probably just add a note to that end, mentioning that if the rights on the file are to permissive, they could get an error.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+**Storage:** 256GB
+
+**Open Ports:** SSH 22, Custom TCP 9090
+
+**Notes:** Enable public IPv4
+
+### Connect to the System
+
+Example command to ssh to the system with a private key
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Is there an online howto about how to get the private key for an EC2 instance? If so, I would tend to link to it here.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+```bash
+chmod 600 ***key***.pem
+ssh -i ***key***.pem ubuntu@***.compute.amazonaws.com
+```
+
+Update packages you can install
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: I would add something like: "Continuing in the ssh shell on the EC2 machine..." just to be 100% clear we are working there now.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+```bash 
+sudo apt-get update
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Why is this needed?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+### Move Scorer.zip to EC2
+
+This is the python scoring pipeline you can download from a completed experiment, run the following from the DAI instance or your local machine:
+
+```bash
+scp -i ***key***.pem scorer.zip ubuntu@***.compute.amazonaws.com:
+```
+
+```
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: missing `bash`

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+**Notes:** Enable public IPv4
+
+### Connect to the System
+
+Example command to ssh to the system with a private key
+
+```bash
+chmod 600 ***key***.pem
+ssh -i ***key***.pem ubuntu@***.compute.amazonaws.com
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Since we are using the machine name a lot, would it make sense to ask the user to store it in an environment variable? Then the user could simply copy&paste the commands here instead of having to rewrite all of them. Wdyt?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+```bash
+unzip scorer.zip
+cd scoring-pipeline/
+```
+
+### Set the License Key 
+
+```bash
+export DRIVERLESS_AI_LICENSE_KEY="oLqLZXMI0y..."
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Should we tell them where to get it or what it is? Or is that obvious?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+
+
+## Install the Requirements
+
+### Python 3.6
+
+```bash
+sudo apt install python3.6 python3.6-dev python3-pip python3-dev python-virtualenv python3-virtualenv
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Should this say `apt-get`?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+### OpenBLAS
+
+```bash
+sudo apt-get install libopenblas-dev
+```
+
+### Upzip
+
+```bash
+sudo apt-get install unzip
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: We have just used `unzip` above. Should we move this section higher?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+```bash
+virtualenv -p python3.6 env
+source env/bin/activate
+pip install -r requirements.txt
+```
+
+The following are okay!
+
+![image-20191119163130015](/Users/mtanco/Library/Application Support/typora-user-images/image-20191119163130015.png)
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Wrong path, please include the image in the PR and refer to it by a local path :+1:.

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+### Test Install
+
+```bash
+python example.py
+```
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Should we mention the expected output/result of this?

File: python-scoring-pipeline/http_ec2_ubuntu.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+Kick off the HTTP Server
+
+```bash
+bash run_http_server.sh
+```
+
+### Get Prediction for a Row:
+
+Change localhost to ***.compute.amazonaws.com to run from your local machine
```
Comment (osery) [inline python-scoring-pipeline/http_ec2_ubuntu.md]: Maybe it would help if we were consistent throughout the doc and either use `localhost` or `***.compute.amazonaws.com` everywhere. Or maybe even better, ask the user to set a variable `SCORER_HOST` and use that consistently allowing them to just copy&paste the commands as they go (instead of editing each of them).

File: python-scoring-pipeline/http_ec2_ubuntu.md (hunk 3 above)
Comment (mtanco) [reply python-scoring-pipeline/http_ec2_ubuntu.md]: I will rewrite this! I specifically meant that the Python Scoring will work for any linux but that the commands were only for/tested with Ubuntu.

File: python-scoring-pipeline/http_ec2_ubuntu.md (hunk 9 above)
Comment (mtanco) [reply python-scoring-pipeline/http_ec2_ubuntu.md]: This is a great idea

**Comment by mtanco** [comment]
@osery Thanks for the amazing feedback, took all into consideration and made all changes. None of the existing folders are mentioned in the read me - is there any specific format that's desired? Happy to update it for all folders but want to check first

**Comment by mtanco** [comment]
I will make a separate PR for the README for all the folders
